File: go/tests/kids_board_test.go
// kids_board_test.go
//
// Verifies the CircleAI.Kids port (kids_board.go): content by age band
// (ordered), limits set/get, usage-today totals per kind, and over-limit
// detection against screen/reading caps.

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestKids_ContentAndLimits(t *testing.T) {
	b := circleai.NewInMemoryKidsBoard()
	b.AddContent(circleai.KidsContent{ContentId: "c2", Title: "Zoo Tales", AgeBand: circleai.AgeAppropriatenessPreschool, Kind: "video", Tags: []string{"animals"}})
	b.AddContent(circleai.KidsContent{ContentId: "c1", Title: "ABC Song", AgeBand: circleai.AgeAppropriatenessPreschool, Kind: "video", Tags: []string{"letters"}})
	b.AddContent(circleai.KidsContent{ContentId: "c3", Title: "Algebra", AgeBand: circleai.AgeAppropriatenessTeen, Kind: "course"})

	pre := b.ContentFor(circleai.AgeAppropriatenessPreschool)
	if len(pre) != 2 || pre[0].Title != "ABC Song" || pre[1].Title != "Zoo Tales" {
		t.Fatalf("content-for band (ordered by Title) failed: %+v", pre)
	}

	b.SetLimits(circleai.DailyTime{KidName: "Kid", ScreenLimit: 2 * time.Hour, ReadingLimit: time.Hour})
	if got, ok := b.LimitsFor("Kid"); !ok || got.ScreenLimit != 2*time.Hour {
		t.Fatalf("limits-for = %+v ok=%v", got, ok)
	}
	if _, ok := b.LimitsFor("Nobody"); ok {
		t.Fatalf("limits for unknown kid must be false")
	}
}

func TestKids_UsageAndOverLimit(t *testing.T) {
	b := circleai.NewInMemoryKidsBoard()
	now := time.Date(2026, 7, 10, 15, 0, 0, 0, time.UTC)
	b.SetLimits(circleai.DailyTime{KidName: "Kid", ScreenLimit: 2 * time.Hour, ReadingLimit: time.Hour})
	b.RecordTime(circleai.TimeLog{KidName: "Kid", Kind: "screen", Duration: 90 * time.Minute, AtUtc: now})
	b.RecordTime(circleai.TimeLog{KidName: "Kid", Kind: "screen", Duration: 45 * time.Minute, AtUtc: now.Add(time.Hour)})
	b.RecordTime(circleai.TimeLog{KidName: "Kid", Kind: "screen", Duration: 30 * time.Minute, AtUtc: now.AddDate(0, 0, -1)}) // yesterday

	if used := b.UsedToday("Kid", "screen", now); used != 135*time.Minute {
		t.Fatalf("used today = %v, want 2h15m", used)
	}
	// 135m > 120m screen cap -> over.
	if !b.OverLimit("Kid", "screen", now) {
		t.Fatalf("expected over screen limit")
	}
	// No reading logged -> under.
	if b.OverLimit("Kid", "reading", now) {
		t.Fatalf("expected under reading limit")
	}
	// Unknown kind -> never over (C# TimeSpan.MaxValue branch).
	if b.OverLimit("Kid", "music", now) {
		t.Fatalf("unknown kind must never be over limit")
	}
	// Unknown kid -> false.
	if b.OverLimit("Nobody", "screen", now) {
		t.Fatalf("unknown kid must not be over limit")
	}
}

File: typescript/tests/kids_board.test.ts
// kids_board.test.ts
// Verifies the CircleAI.Kids port: age-banded content sorted by title, used-today
// (same UTC day) + over-limit for screen/reading (and no cap for other kinds).

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryKidsBoard,
  KidsDomainContext,
  AgeAppropriateness,
  kidsContent,
  dailyTime,
  timeLog,
} from "../src/kids/index";

const MIN = 60_000; // one minute in ms

describe("InMemoryKidsBoard", () => {
  it("lists content in a band sorted by title", () => {
    const b = new InMemoryKidsBoard();
    b.addContent(kidsContent("c1", "Zebra", AgeAppropriateness.Toddler, "video", ["animals"]));
    b.addContent(kidsContent("c2", "Apple", AgeAppropriateness.Toddler, "book", ["food"]));
    b.addContent(kidsContent("c3", "Space", AgeAppropriateness.Teen, "game", ["science"]));
    assert.deepEqual(
      b.contentFor(AgeAppropriateness.Toddler).map((c) => c.contentId),
      ["c2", "c1"],
    );
    assert.equal(AgeAppropriateness.Toddler, 0);
    assert.equal(AgeAppropriateness.Teen, 5);
  });

  it("sums used-today only for the same UTC calendar day and matching kind", () => {
    const b = new InMemoryKidsBoard();
    const now = new Date("2026-01-10T20:00:00Z");
    b.recordTime(timeLog("Kai", "screen", 30 * MIN, new Date("2026-01-10T08:00:00Z")));
    b.recordTime(timeLog("Kai", "screen", 15 * MIN, new Date("2026-01-10T18:00:00Z")));
    b.recordTime(timeLog("Kai", "screen", 60 * MIN, new Date("2026-01-09T18:00:00Z"))); // yesterday
    b.recordTime(timeLog("Kai", "reading", 10 * MIN, new Date("2026-01-10T09:00:00Z")));
    assert.equal(b.usedToday("Kai", "screen", now), 45 * MIN);
    assert.equal(b.usedToday("Kai", "reading", now), 10 * MIN);
  });

  it("flags over-limit for screen/reading, false when no limits, no cap for other kinds", () => {
    const b = new InMemoryKidsBoard();
    const now = new Date("2026-01-10T20:00:00Z");
    // No limits set yet.
    b.recordTime(timeLog("Kai", "screen", 120 * MIN, new Date("2026-01-10T08:00:00Z")));
    assert.equal(b.overLimit("Kai", "screen", now), false);

    b.setLimits(dailyTime("Kai", 60 * MIN, 30 * MIN));
    assert.equal(b.limitsFor("Kai")?.screenLimitMs, 60 * MIN);
    assert.equal(b.overLimit("Kai", "screen", now), true); // 120 > 60
    // C# OverLimit selects the cap case-insensitively, but the usage total comes
    // from UsedToday, which matches Kind with an ordinal (case-sensitive) `==`
    // (KidsPrimitives.cs line 43). "SCREEN" therefore matches no "screen" logs →
    // used = 0 → 0 > 60min is false. The kind casing only affects the cap, not
    // which logs are summed.
    assert.equal(b.overLimit("Kai", "SCREEN", now), false);

    b.recordTime(timeLog("Kai", "reading", 10 * MIN, new Date("2026-01-10T09:00:00Z")));
    assert.equal(b.overLimit("Kai", "reading", now), false); // 10 <= 30

    // Unknown kind → no cap (TimeSpan.MaxValue), never over.
    b.recordTime(timeLog("Kai", "outdoor", 999 * MIN, new Date("2026-01-10T09:00:00Z")));
    assert.equal(b.overLimit("Kai", "outdoor", now), false);
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(KidsDomainContext.systemPromptSnippet.includes("[DOMAIN: Kids]"));
    assert.deepEqual(KidsDomainContext.complianceFlags, [
      "POPIA_Childrens_Data",
      "COPPA_principles",
      "Childrens_Act",
      "CAPS_curriculum",
    ]);
    assert.deepEqual(KidsDomainContext.suggestedTools, ["educational_content", "story_tools", "quiz_tools"]);
  });
});
